ports/rp2/msc_disk.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020-2021 Damien P. George
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#include "tusb.h"
+#if CFG_TUD_MSC
+#include "mpconfigboard.h"
+#include "hardware/flash.h"
+#include "hardware/sync.h"
+#include "pico/stdlib.h"
+#include "py/mphal.h"
+
+// This implementation does Not support Flash sector caching.
+#if MICROPY_FATFS_MAX_SS != FLASH_SECTOR_SIZE
+#error MICROPY_FATFS_MAX_SS must be the same size as FLASH_SECTOR_SIZE
+#endif
+
+#define BLOCK_SIZE          (FLASH_SECTOR_SIZE)
+#define BLOCK_COUNT         (MICROPY_HW_FLASH_STORAGE_BYTES / BLOCK_SIZE)
+#define FLASH_BASE_ADDR     (PICO_FLASH_SIZE_BYTES - MICROPY_HW_FLASH_STORAGE_BYTES)
+#define FLASH_MMAP_ADDR     (XIP_BASE + FLASH_BASE_ADDR)
+
+static bool ejected = false;
+
+bool tud_msc_device_ejected() {
+    return ejected;
+}
+
+MP_WEAK int tud_msc_device_ejected_cb(bool ejected) {
+    (void)ejected;
+    return 0;
+}
+
+// Invoked when received SCSI_CMD_INQUIRY
+// Application fill vendor id, product id and revision with string up to 8, 16, 4 characters respectively
+void tud_msc_inquiry_cb(uint8_t lun, uint8_t vendor_id[8], uint8_t product_id[16], uint8_t product_rev[4]) {
```

Comment:
On Linux I'm seeing the following in dmesg:
```
[193765.844070] scsi host1: scsi scan: INQUIRY result too short (5), using 36
```
Not sure why...

Reply:
I'm not sure what that means, I followed TinyUSB MSC example/templates, I could look into it.

Reply:
I can't find more info on this, it should be returning 36 bytes here https://github.com/hathach/tinyusb/blob/master/src/class/msc/msc_device.c#L751 note `sizeof(scsi_inquiry_resp_t)` is 36 bytes. I searched TinyUSB repo there are no related issues, however I've seen this error in dmesg logs of unrelated issues and it seems to be ignored.

Reply:
I checked other TinyUSB MSC implementations and did not see this message, so it must be something to do with the implementation here.  I did look into it a bit but could not find the problem.

Reply:
I tested CircuitPython on Nano RP2040, and I see the same warning, so I think it's something in the upper layers (in TinyUSB).

Reply:
If I understood correctly, the response size of all SCSI inquiry commands should be 36 bytes, in SCSI-2 standard and higher, so it seems that TinyUSB is maybe using a legacy variable sized response. It seems like a harmless warning, but either way it's definitely something that must be fixed upstream.

Reply:
I tested CircuitPython on ESP32-S2 and the warning was not there.

Also, the RP2040 bootloader does not have this issue, and AFAIK that's based on TinyUSB??

So it's strange, where does the issue lie?

Reply:
I found the issue and will send a fix upstream, the issue is the "additional length" field is set to 0, instead of the remaining inquiry command response data (size - 4). The Linux driver warning ignores it and assumes its length is the minimum response (36 bytes). Here's the dmesg after the fix:
```bash
53812.065278] usb 4-2: new full-speed USB device number 120 using xhci_hcd
[53812.231281] usb 4-2: New USB device found, idVendor=2341, idProduct=015e, bcdDevice= 1.00
[53812.231294] usb 4-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[53812.231298] usb 4-2: Product: Board in FS mode
[53812.231301] usb 4-2: Manufacturer: MicroPython
[53812.231304] usb 4-2: SerialNumber: 50234274c853861c
[53812.257376] cdc_acm 4-2:1.0: ttyACM0: USB ACM device
[53812.260034] usb-storage 4-2:1.2: USB Mass Storage device detected
[53812.260334] scsi host0: usb-storage 4-2:1.2
[53813.287534] scsi 0:0:0:0: Direct-Access     Micropy  Mass Storage     1.0  PQ: 0 ANSI: 2
[53813.288426] sd 0:0:0:0: [sda] 2048 4096-byte logical blocks: (8.39 MB/8.00 MiB)
[53813.289011] sd 0:0:0:0: [sda] Write Protect is off
[53813.289015] sd 0:0:0:0: [sda] Mode Sense: 03 00 00 00
[53813.291011] sd 0:0:0:0: [sda] No Caching mode page found
[53813.291013] sd 0:0:0:0: [sda] Assuming drive cache: write through
[53813.299579]  sda: sda1
[53813.302998] sd 0:0:0:0: [sda] Attached SCSI removable disk
```

> I tested CircuitPython on ESP32-S2 and the warning was not there.

It should be there, because it's not fixed upstream. Note it may be getting random values from the stack.

Reply:
> Also, the RP2040 bootloader does not have this issue, and AFAIK that's based on TinyUSB??
I could be wrong but isn't the bootloader is in ROM ?  I don't think it's TinyUSB based, or maybe they maintain their own fork.

Reply:
Good work solving it!

Re RP2040 bootloader:  source is https://github.com/raspberrypi/pico-bootrom and it looks to be a custom USB MSC implementation.

Reply:
@dpgeorge https://github.com/hathach/tinyusb/pull/1349

Reply:
FYI: https://github.com/raspberrypi/pico-bootrom/blob/7a751f28436a8ee4849ac7799eb3aa1457e6f091/usb_device_tiny/scsi_ir.h#L32